### PR TITLE
Poll services

### PR DESCRIPTION
**Pull request by SeavantUUz**
implement a new mechanism to manage blocklog-series processor

### CODE REVIEW

File: app/plugins/blocklog_fast_foward_manager_service.go [hunk 1]
```diff
@@ -31,15 +31,15 @@ func (cp FastForwardManagerCheckpoint) HasNeedSyncProcessors() bool {
...
 	var progresses []*Progress
-	cp.db.Where(&Progress{FastForward:true}).Find(progresses)
+	cp.db.Where(&Progress{FastForward:true}).Find(&progresses)
 	return progresses
 }
 
 func (cp FastForwardManagerCheckpoint) TryToTransferProcessorManager(progress *Progress) error {
 	blogLog := &iservices.BlockLogRecord{}
-	cp.db.Where(&iservices.BlockLogRecord{}).Last(blogLog)
+	cp.db.Last(blogLog)
```
Comment (dai-shuo) [inline app/plugins/blocklog_fast_foward_manager_service.go]: this won't work with the customized table name of block logs.

File: app/plugins/fast_foward_manager_service.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"github.com/coschain/contentos-go/iservices"
+	"github.com/jinzhu/gorm"
+	"github.com/sirupsen/logrus"
+)
+
+type FastForwardManagerService struct {
+	*ForwardManagerService
+}
```
Comment (dai-shuo) [inline app/plugins/fast_foward_manager_service.go]: FastForwardManagerService & SyncForwardManagerService are just aliases of ForwardManagerService.
No necessity to define them at all. Though it's not a bug.

File: app/plugins/forward_manager_service.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+		if atomic.LoadInt32(&s.stop) != 0 {
+			userBreak = true
+			break
+		}
+		processor, ok := s.mainProcessors[progress.Processor]
+		if !ok {
+			continue
+		}
+		minBlockNum, maxBlockNum := progress.BlockHeight+1, progress.BlockHeight+maxJobSize
+		for blockNum := minBlockNum; blockNum <= maxBlockNum; blockNum++ {
```
Comment (dai-shuo) [inline app/plugins/forward_manager_service.go]: This is a ***downgrade*** of consistency.
According to my understanding, global consistency is not guaranteed any more. Instead, consistency is now in a group level. Tables are consistent only if they're maintained by the same `IBlockLogProcessor`.
Query results may be inaccurate if we need to join tables from different groups, e.g. holders & ecosys_reward.
We should discuss whether this is acceptable.

**Review by dai-shuo — CHANGES_REQUESTED** [review]

File: app/plugins/forward_manager_service.go [hunk 4]
```diff
@@ -29,32 +28,12 @@ type ForwardManagerService struct {
 	point Checkpoint
```
Comment (dai-shuo) [inline app/plugins/forward_manager_service.go]: `point` and `Checkpoint` in no use now

**Review by dai-shuo — CHANGES_REQUESTED** [review]
keep "blocklog_process" table. there're dependent services.

File: app/plugins/forward_manager_service.go (hunk 4 above)
Comment (SeavantUUz) [reply app/plugins/forward_manager_service.go]: removed unused checkpoint now

**Comment by SeavantUUz** [comment]
> keep "blocklog_process" table. there're dependent services.

File: app/plugins/blocklog_fast_foward_manager_service.go (hunk 1 above)
Comment (dai-shuo) [reply app/plugins/blocklog_fast_foward_manager_service.go]: fixed by 7ddac6b

File: app/plugins/forward_manager_service.go (hunk 3 above)
Comment (dai-shuo) [reply app/plugins/forward_manager_service.go]: fixed by 7ddac6b

**Pull request merged**